command/agent/exec/exec.go

```diff
@@ -0,0 +1,260 @@
+		config:    cfg,
+		procState: processStateNotStarted,
+	}
+
+	return &server
+}
+
+func (s *Server) Run(ctx context.Context, incomingVaultToken chan string) error {
+	latestToken := new(string)
+	s.logger.Info("starting exec server")
+	defer func() {
+		s.logger.Info("exec server stopped")
+	}()
+
+	if len(s.config.AgentConfig.EnvTemplates) == 0 || s.config.AgentConfig.Exec == nil {
+		s.logger.Info("no env templates or exec config, exiting")
+		return nil
+	}
+
+	managerConfig := ctmanager.ManagerConfig{
+		AgentConfig: s.config.AgentConfig,
+		Namespace:   s.config.Namespace,
+		LogLevel:    s.config.LogLevel,
+		LogWriter:   s.config.LogWriter,
+	}
+
+	runnerConfig, err := ctmanager.NewConfig(managerConfig, s.config.AgentConfig.EnvTemplates)
+	if err != nil {
+		return fmt.Errorf("template server failed to generate runner config: %w", err)
+	}
+
+	// We leave this in "dry" mode, as there are no files to render;
+	// we will get the environment variables rendered contents from the incoming events
+	s.runner, err = manager.NewRunner(runnerConfig, true)
+	if err != nil {
+		return fmt.Errorf("template server failed to create: %w", err)
+	}
+
+	s.numberOfTemplates = len(s.runner.TemplateConfigMapping())
+
+	for {
+		select {
+		case <-ctx.Done():
+			s.runner.Stop()
+			if s.proc != nil {
+				s.proc.Stop()
+			}
+			s.procState = processStateStopped
+			return nil
+		case token := <-incomingVaultToken:
+			if token != *latestToken {
+				s.logger.Info("exec server received new token")
+
+				s.runner.Stop()
+				*latestToken = token
+				newTokenConfig := ctconfig.Config{
+					Vault: &ctconfig.VaultConfig{
+						Token:           latestToken,
+						ClientUserAgent: pointerutil.StringPtr(useragent.AgentTemplatingString()),
+					},
+				}
+
+				// got a new auth token, merge it in with the existing config
+				runnerConfig = runnerConfig.Merge(&newTokenConfig)
+				s.runner, err = manager.NewRunner(runnerConfig, true)
+				if err != nil {
+					s.logger.Error("template server failed with new Vault token", "error", err)
+					continue
+				}
+				go s.runner.Start()
+			}
+
+		case err := <-s.runner.ErrCh:
+			s.logger.Error("template server error", "error", err.Error())
+			s.runner.StopImmediately()
+
+			// Return after stopping the runner if exit on retry failure was specified
+			if s.config.AgentConfig.TemplateConfig != nil && s.config.AgentConfig.TemplateConfig.ExitOnRetryFailure {
+				return fmt.Errorf("template server: %w", err)
+			}
+
+			s.runner, err = manager.NewRunner(runnerConfig, true)
+			if err != nil {
+				return fmt.Errorf("template server failed to create: %w", err)
+			}
+			go s.runner.Start()
+		case <-s.runner.TemplateRenderedCh():
+			// A template has been rendered, figure out what to do
+			s.logger.Debug("template rendered")
+			events := s.runner.RenderEvents()
+
+			// events are keyed by template ID, and can be matched up to the id's from
+			// the lookupMap
+			if len(events) < s.numberOfTemplates {
```

Comment:
Will it always make sense for us to wait to render all templates? For static secrets only, it probably does, as they're the same cadence, but maybe my template has a dynamic secret with a TTL of 1d, and another with a TTL of 1h. Will the 1h TTL secret be waiting for a day every time?

It might also be that I'm misinterpreting this :)

Reply:
Very good point. Maybe a better approach would be to have a ~1 min window of waiting for the last update before restarting. Alternatively just restart on every template change since it's not likely that secrets all change at once. Even if they do, the restart takes some time and by the time it is done, likely all the other secrets have been updated.

Reply:
Also just noticed that the comment needs to be updated

Reply:
will update the comment

> Maybe a better approach would be to have a ~1 min window of waiting for the last update before restarting.

sounds like a debounce function of some sort, have thought about it but don't think its a core feature

also worth noting, there's no way to tell if the secret being compiled is dynamic or static (at least at the time of writing this)

Reply:
Can you foresee any issue if we don't wait for all templates, i.e. call `bounce` upon a single event? Based on our discussion here, I think that might make sense, so that templates aren't waiting for the slowest one to be rendered.

Actually, thinking about it more, there could be a bug with the existing approach too -- given a 1hr TTL secret and 24hr TTL secret, we'd be restarting every two hours, since we'd get two events from the 1hr TTL secret, right?

Reply:
I discussed this thread with Daniel on Slack -- I misinterpreted the events as being wiped after this executes, for some reason. 

I'm going to resolve this conversation, but for prosperity -- this check will be true both after we've finished rendering the initial set of templates, and for every consecutive re-render